### PR TITLE
use Makefile for all tasks

### PR DESCRIPTION
**Pull request by merlinran**
It is based off https://github.com/Twisted-Fields/acorn-precision-farming-rover/pull/12 and the change specific to this PR is in a single commit https://github.com/Twisted-Fields/acorn-precision-farming-rover/commit/541629ae9eacfe52568334ede8a3e94a2fd3200d

### CODE REVIEW

**Comment by tlalexander** [comment]
Make has always seemed pretty hard to follow for me. Are there any alternatives you can think of?

Is it just executing some commands? I am curious if we could achieve much the same results with a python script?

could call it acorn-run.py and mark the execute bit, so instead of calling
make simulation
we could call
acorn-run.py simulation

Or is make really needed? Let me know your thoughts. Thank you!

**Comment by merlinran** [comment]
Some sort of build automation seems useful to, for example, rebuild the docker image automatically before running simulation if any of the `requirements.txt` gets changed, or some other guy updating to the latest code would find it fail to run because of missing Python packages, and spend some minutes to figure out why and do the docker build by themselves. We could do the check of timestamp and rebuilding in Unix shell, but that's less structural than a tool dedicated to this, and IMHO shell is messier and harder to debug than Makefile (the errors you encountered when running `make` [in the other PR](https://github.com/Twisted-Fields/acorn-precision-farming-rover/pull/18#issuecomment-1004525251) are due to differences among shell flavours). If we use Python, there would be tons of `os.system`/`subprocess.popen` to run the commands and get the output, which seems even more suboptimal.

I chose `make` because of its ubiquity - and many developers are more or less familiar with it. An alternative would be https://www.scons.org which is Python-like, but that is yet another domain specific language for most of the people to learn. Same for https://github.com/bazelbuild/bazel and others.

From my experience, `Makefile` seldom changes once gets established, and though the syntax has many dark corners I myself would frustrate too, the basic rules are simple. Actually the majority of the `Makefile` are plain shell scripts. I'll comment on the `Makefile` features. Lmk if they make sense to you and if not, we can discuss on the alternatives.

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,64 @@
+LOCAL_IMAGE = acorn_docker:1.0
+REMOTE_IMAGE = merlinran/acorn_docker
```
Comment (merlinran) [inline Makefile:2]: Variables are similar to shell variables, you just need to reference them via `$(var)`, instead of `$${var}`(`${var}` in normal shell, but since we are running shell scripts within `Makefile`, we need to escape `$`, similar to why we have to use `\\` to represent `\` in C string).

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+.PHONY: list
```
Comment (merlinran) [inline Makefile:4]: `. PHONY` is to tell `Make` that the target is not a real file. Without it, if there's a file called `list` in the directory, make would think: the file is already there and up-to-date, I won't do anything.

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+list:
```
Comment (merlinran) [inline Makefile:5]: Since this is the first target in the file, `make` without a target simply runs this rule. And this rule has no dependencies.

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	@echo "Welcome! You may want to try one of the following:\n---"; \
```
Comment (merlinran) [inline Makefile:6]: `@` tells `make` to not prompt the command follows. The rest are just normal shell scripts. This rule is a trick to list the main targets from the `Makefile` itself.

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	grep "^.PHONY:" Makefile | grep "#" | cut -d ":" -f 2- | sed "s/\w*#/\t#/g" | sed "s/^/make/"
+
+.PHONY: simulation # Run the vehicle and server containers in simulation mode.
+simulation: docker-image
```
Comment (merlinran) [inline Makefile:10]: Here `docker-image` is a dependency of `simulation`. Before running the body of the rule, `make` first makes sure `docker-image` is up-to-date, by recursively checking the rule at [L59](https://github.com/Twisted-Fields/acorn-precision-farming-rover/pull/17/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R59) and running the body if needed.
In this case, the dependency chain is `simulation` <- `docker-image` <- `Dockerfile` <- `vehicle/requirements.txt` and `server/requirements.txt`. As long as the upstream target is newer than the downstream target, the rule body will be run. It would be very difficult to express such relationship in plain shell or Python.

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	@DOCKER_COMPOSE_FILE=docker-compose-simulation.yml make restart-docker-compose &&\
+	echo "Please visit http://localhost"
+
+.PHONY: attach-vehicle # Attach to the shell of the vehicle container. It creates the container if it doesn't exist.
+attach-vehicle:
+	@test -z `docker ps -f "name=acorn_vehicle" -q` && make docker-vehicle; \
```
Comment (merlinran) [inline Makefile:16]: Here's again plain shell. Start the container if it doesn't exist, then `docker exec` it. `test -z` returns `0`(success) if the output of the command is empty, in which case `make docker-vehicle` will run; Otherwise it returns `1`(fail), and the `&&` short-circuits. The command after `;` is not affected, hence being run unconditionally.

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+.PHONY: docker-test # Start the vehicle container in test mode and run the tests in it.
+docker-test: docker-image
+	@docker-compose -f docker-compose-test.yml up --remove-orphans -d && \
+	docker exec -it acorn_vehicle make test
+
+.PHONY: push-image # Build and push image to Docker Hub to be used by CI.
+push-image: docker-image
+ifeq ($(shell uname -m),arm64)
+	docker buildx build -t $(REMOTE_IMAGE) --platform linux/amd64,linux/arm64 --push .
```
Comment (merlinran) [inline Makefile:36]: This is to allow me building and pushing the image for `linux/amd64` while on my MacBook with M1 processor. Otherwise it just hits the next branch.

File: Makefile [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+else
+	docker build -t $(REMOTE_IMAGE) . && docker push $(REMOTE_IMAGE)
+endif
+
+.PHONY: test  # Run tests on Linux (if the Python dependencies are installed) or inside Docker. Otherwise, you probably want to try `make docker-test` instead.
+test:
+	coverage run -m pytest --log-cli-level DEBUG && coverage report --skip-covered --skip-empty
```
Comment (merlinran) [inline Makefile:43]: This is supposed to be run inside the container, but since you are on Linux, you can also `make test` directly if all required Python packages are installed.

File: Makefile (hunk 4 above)
Comment (merlinran) [reply Makefile:6]: One caveat is to always use tab instead of whitespaces in the rule body, or `make` won't recognize it. Editor may warn you or correct for you if that happens.

**Review by merlinran — COMMENTED** [review]
I think this is basically it. I never dedicated time to learn `make`, just followed the `Makefile`s I had to make changes and checked the manual for things I didn't quite understand.
